### PR TITLE
Ensure all songs and albums have an artist

### PR DESCRIPTION
**Pull request by DaveTBlake**
Until now it has been possible to have songs and albums in the music library that don't have an artist. It is the result of scanning music files that do not have an ARTIST/TPE1 tag. Many users will have artists for every song because they have tagged their music fully, but missing artist can still happen accidentally.

There are 2 issues with this:

1) All queries wanting song or album and their artist(s) have to be left joins. Unfortunately there is a weakness in the SQLite optimiser that means that left joins onto views are not optimised and run horribly slowly. The work around currently implemented is to replace all the views in left joins with the explicit fields and tables, but it is not as easy to maintain or read. See #9081 for more discussion.

2) Songs without artists are hidden in the UI. They do not get listed in a Genre -> Artist -> Album -> song approach. Only the songs node, listing all the songs, would show them somewhere in the potentially 1000s of other songs.

From a data view point in makes a lot of sense for Kodi to collect all songs without an artist tag under a single invented  "missing tag" artist. This would mean that we can use more efficient inner joins in queries, but also that the "missing tag" artist would appear in the artist node and reveal those previously lost songs to the user, encouraging them to correct the tagging.

A verbose discussion is on the forum http://forum.kodi.tv/showthread.php?tid=260165
Following that I have adopted the approach that the invented "missing tag" artist  being unique and different from any other artist that could be scanned from music file tags, using a fake musicbrainz id. The name of the "missing tag" artist displayed by Kodi comes from the **current** language settings, the artist table being updated every scan of a file with a missing artist tag, and on library update.

While there I have improved the indexes on song_artist table.

@phate89 and @Razzeee as follow up from previous PR. 
@Tolriq I have not made any attempt to extend the JSON API with this PR. I am still waiting for feedback from @Montellese on earlier work, so perhaps it is better to do the API changes in a separate PR? I understand that the API needs to expose which artist is the invented "missing tag" one so that consumers can treat it in a special way if they want, but despite discussion I am unclear on the best way to do this. As it is where previously API was returning empty "artist" and "artistid" fields it will return values for both mixed in with all the other artists, so as it is this does not break the API.

### CODE REVIEW

**Comment by Tolriq** [comment]
@DaveTBlake see my answer on the forum I forget to validate previously.

There's no need to change API, but some solutions needs implementation in this PR. as not doing it now will forbid them later depending on choice.

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -309,7 +311,7 @@ void CMusicDatabase::CreateViews()
               "        album.iUserrating, "
               "        album.iVotes, "
               "        bCompilation, "
-              "        (SELECT MIN(song.iTimesPlayed) FROM song WHERE song.idAlbum = album.idAlbum) AS iTimesPlayed, "
+              "        (SELECT AVG(song.iTimesPlayed) FROM song WHERE song.idAlbum = album.idAlbum) AS iTimesPlayed, "
```
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: This seems unrelated and might have snuck in?

Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Yes, it snucked!
OTOH I do prefer AVG to MIN, and it seems a shame to have a DB version step for a single view change (no updating just need the views recreated as they are with a new db file).

Dare I leave it?

Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Sure, but you might want to move it to a separate commit then

**Comment by Montellese** [comment]
I would split the commits up so that the LEFT JOIN optimisation is in its own commit.

Am I right in assuming that if I scan my music library with Kodi in English and then switch to German it will still show "[Missing]" until I rescan my music library?

**Comment by DaveTBlake** [comment]
>Am I right in assuming that if I scan my music library with Kodi in English and then switch to German it will still show "[Missing]" until I rescan my music library?

Yes, until you attempt a library update, a complete rescan is not needed. Or whenever you scan a file without an artist tag. I did think about checking every startup, but not sure where to put such a process and was diswaded on IRC.

While you're here @Montellese any change of final feedback on my music API WIP?

**Comment by Tolriq** [comment]
@DaveTBlake this translation problem whatever automated solution on Kodi side chosen will never handle the API consumer point of view ;) 
(Kodi in German remote phone in English, data displayed in German while user want it in English)
That's the whole point of my first message about VA translations way before you started working on this PR.

API : Pure data, Kodi settings agnostic, I fell like repeating same stuff over and over ;)

But let's continue this on forum to avoid spam here.

**Comment by DaveTBlake** [comment]
>This translation problem whatever automated solution on Kodi side chosen will never handle the API consumer point of view

It is not trying to do so, how could it? Translation is obviously best done at the UI end just as you suggest.

All I am saying is that the database will hold language setting dependant name for the missing tag artist. I do not advise we use that name to identify the artist for processing of any kind. All I do is suggest that unless the skin or API consumer has decided to use something else we provide that name as a default.

Unlike VA where you end up with multiple artists this will only allow one

**Comment by Tolriq** [comment]
Since you answer here ;)

Yes translation is best done at UI level and this is the whole story. 
Kodi support profiles with different language but shared medias info and sources.

This means that with translation in DB and not handled at the DB wrapper level or GUI level, this missing artist displayed language will be at best totally random.

Take 2 profiles in 2 languages, displayed value will depend on who last scanned media. And user will be lost at understanding why this artist change it's name all the time.

As I explained a lot, having an unique artist is good and not related at all to VA that creates multiple artists. This is not the debate about multiple data, this is a debate about how the data is stored and displayed.

I guess my English is really worse than I thought, but since @Montellese have directly pointed to that fact I hope he'll explain better than me.

**Comment by DaveTBlake** [comment]
There is more discussion on the forum, but to summarise for those here I am going to look at how providing translation of "[Missing]" into the live language version in the middle layer tween DB and UI impacts timing. 

I get that the API generally wants to be independant of Kodi settings, but do some API consumers want to receive the same translated artist name as the Kodi UI?

**Comment by DaveTBlake** [comment]
Ok, I have split the commits as suggested (thanks for the dummy's guide @Razzeee )
I have also embraced the fact that varying the db content with language setting was not the best design. Instead translation of the artist name to the language of the current profile takes place on the way into CFileItem.MusicInfoTag. Hence @Montellese it does not matter what the settings are when the music is scanned, the user will see the current language.

The missing tag artist name only appears on the artists node. While a song is playing or listed in song node if it has no artist tags them no artist name is shown. This is rather than clutter the display with [Missing], when they have maybe put the artist name in the song title or filename, something @Millencolin007 asked for.

Hence for songs without an artist tag AudioLibrary.GetSongs will return "artist" = [""], and "artistid" = [1895] (whatever the id). But AudioLibrary.GetArtists will return "artistid" = [1895], and  "artist" = "[Missing]" (current language). It is this way for GUI, and doing something special for API would be a right fiddle. If we have a GetMissingArtistID API routine I am hoping this will be workable. But yet to add anything for the API.

The other design change discussed on the forum was to use a fixed/known artist id e.g. 1 rather than whatever we get when it is needed as currently implemented. Anyone other than @Tolriq have a view on this?

**Comment by Tolriq** [comment]
For those who do not read the forum, I still think JSON API should return the data in English and not depend on Kodi settings.

And if no fixed ID then GetMissingArtistID is mandatory to allow remotes to show this value in correct language independent of Kodi one.

But since CMusicDatabase is recreated on most access having also internals to query for this ID so often does tends to push towards a fixed ID :( Is there's another place to store the resolved value ?

And since I'm still lost in music handling, what will be the value of display artist ?

**Comment by zag2me** [comment]
After a quick chat on slack, this seems fine to me. Assuming we don't use a real mbid there should be no problems with metadata lookups on those missing artists.

**Comment by DaveTBlake** [comment]
Both API and GUI gets artist data from CFilItem.MusicInfoTag, and translation into the profile language takes place into this - at the point when the artistID is known, as MusicInfoTag does not store ID so it would be harder to implement translation on the way out of MusicInfoTag. Consequently to always give the API "missing artist tag" in English but the GUI a translated value would mean some kind of flag and test when MusicInfoTag is populated. It is possible, sure, but seems an extra complication especially when the API consumer can use the ID to do what they want with the name (now always in English) anyway.

So @Tolriq is clear on wanting this name in English, and I know that @Razzeee on IRC has said he thinks the opposite, API gets text in the profile language. Can we have that on the record here @Razzeee ? What is the consensus?

I also think if we have a "magic number" approach, a fixed and well know ID for the missing tag artist, we would still need a GetMissingArtistId API call for those that did not read the documentation. Again, any consensus on this?

>what will be the value of display artist ?

"Display artist" = ""

**Comment by Razzeee** [comment]
Sure, I can say it again. I think this is a different case then the normal "json api should be free to get what it wants" and even if we go for it, I think it's out of scope for the PR?

As I'm not sure if there is an json rpc endpoint to get the translated value of something. So we would have to find a solution there. Translating in a third party tool seems wrong, as that will break sometimes, when the third party app doesn't use the same translation as the core. So it will lead to a bad UX due to bad api design.

**Comment by Tolriq** [comment]
Why is it different ? Why are date independent of Kodi UI settings ? Why are PVR timer week day in English, and everything else ?

Because an API should return data, the fact that Kodi is loggued with a Profile X or Y with different language settings should not return different data.

I totally understand that it's maybe not possible with current code base. But there's no valid reason for an API that can consume data to have data translated depending on Kodi UI. This is nonsense.

Data should be returned in English as it's currently the default, if one day the API is extended to support querying for a specific language then API will return in that language.

Having data that change it's value depending on random settings is not an API.

If I have my phone in English and Kodi in German why would text be displayed in German ?

But as you said if not possible it can be out of the PR if the API consumers can identify this user and do hack on their side (GetMissingArtistId), but do not say that an API should return random data depending on Kodi settings, this is not API in that case :)

> Translating in a third party tool seems wrong, as that will break sometimes, when the third party app doesn't use the same translation as the core. So it will lead to a bad UX due to bad api design.

Bad design is an API that think in behalf of the API consumer... This is clearly not Kodi's business if any API consumer display data translated the way they want.

I will certainly not display [Missing] because I do not like the [] concept, this is not Kodi API job to force me to display what it want to.


About 
> I also think if we have a "magic number" approach, a fixed and well know ID for the missing tag artist, we would still need a GetMissingArtistId API call for those that did not read the documentation

While having the 2 is not a problem, justifying an API because people would not read a doc sounds over protective. They need to read the documentation of the API for everything. Why would this be different ?

**Comment by Montellese** [comment]
I don't have the time to read all the arguments but IMO
* the API needs to be Kodi settings agnostic. The name of the "missing artist" artist should always be the same independent of what profile and/or settings are loaded
* the API either should not return the "missing artist" artist and provide a way to query for albums/songs without artists OR needs to provide information for API consumers to identify which returned artist is the artificial "missing artist"

**Comment by Razzeee** [comment]
@Tolriq and @Montellese are right. Ignore me @DaveTBlake :)

**Comment by DaveTBlake** [comment]
@Montellese do you have an opinion on whether we have a fixed and well known (documented) artistID for the missing tag artist, or, as so far implemented, allow it to be any artistID (or none if the library has no songs with missing artist) but implement a GetMissingArtistID method in the API?

**Comment by Tolriq** [comment]
I know it's hard to pass messages, but since others say the same as me I'll continue :)

Having the possibility to not have an artist returned by GetMissingArtistID, means that a remote needs to query this value each time in the case that any time in the future the missing artist is created.

From efficiency point of view. This artist should be always created. So remotes know that they are talking to Krypton ask this ID then store it, and avoid querying it all the time.

**Comment by DaveTBlake** [comment]
>From efficiency point of view this artist should be always created.

@Tolriq  as I see it the existence and ID for the missing tag artist would only need to be queried when the library contents may have changed. That is not the same as "every time", and I would think library change (because of the general delete/recreate design) would always prompt lots of querying anyway, one more value request seems inconsequential. I see it differently from you, but what do I know? So I would just like some input from someone else.

I am also aware the @Montellese dislikes "magic numbers". Since he has the final say on API matters I would like to know his view on fixed/known artist id for missing before I go off and spend time changing things ready for the API implementation.

**Comment by Tolriq** [comment]
Well my answer did not talked about magic numbers at all ;) I already gave my point of view on that and agreed that a GetMissingArtistID can fit the need for remotes even if less efficient for long term point of view. Artist roleid 1 is also what you call a magic number, you could have add them in the order they come and query for the roleid of artist each time, but you did not do that because it's less efficient. There's still the get roles and everything. But internally you avoid 1 query every time.

Anyway my answer is only about the fact that GetMissingArtistID should always return a value. You create the artist and 2 others roles even if they do not exist, because it's logical to always have those. The missing artist is also a fixed value.

Anyway about your point of view, you seems pissed off by the fact that your attempts to prove I'm wrong only bring answers that confirm what I said and I'm sorry that you feels like that.

**Comment by DaveTBlake** [comment]
>Anyway about your point of view, you seems pissed off by the fact that your attempts to prove I'm wrong only bring answers that confirm what I said and I'm sorry that you feels like that.

@Tolriq where you have made good points, or where we have disagreed but others have confirmed your view I have adapted my implementation (or am in the process of doing so). It has never been about proving you wrong, just reasoned discussion.  I do feel weary, working on Kodi is like being in treacle sometimes, and I am in pain with a health problem which does not help. I truely would prefer collaboration rather than eternal battle.

Regarding a fixed artistId for missing tag artist, I would just like some input from someone else, @Montellese would be great, that this is the implementation the team want. Kodi core does not need it, but maybe we need to do it for the API efficiency. Just I do not want to spend time only to be told later it is wrong, so is it so bad to ask for a second opinion?

**Comment by Tolriq** [comment]
Absolutely not and I'm sorry for your health problem.

But my problem was that I said : "GetMissingArtistID needs to always return value", you answer: "no but I want an opinion from Montellese about magic numbers". 

Those are completely unrelated so this sounded like we started again (as often) one talk about A and other continue to answer about B without reading about A.

My only concern is only about GetMissingArtistID returning a value, meaning just call addMissingArtist during upgrade process. The fact that you change later or not to fixed id have nothing to do with that.

**Comment by DaveTBlake** [comment]
@Tolriq as I understand it your view is that it would be most efficient for the missing tag artist to have a fixed/known artistID. But failing that and the artist ID could vary between Kodi databases, it must always exist (even if the user does not have any songs with missing artist tags) and be returned by API method GetMissingArtistID. And finally the artist name for this artist returned by AudioLibrary.GetArtists and GetArtistDetails must be a fixed/known value (in English). Adequate summary?

Meanwhile I would like a second opinion on the question of using a fixed ID.  If we go that way then the artist always existing becomes a given.  Hence I didn't get into discussion about it until knowing answer to the primary question.

Meanwhile the more we debate it seems the less likely I will ever get that second opinion!

**Comment by Tolriq** [comment]
Text value should be English, fixed or known is no importance correct API usage for remotes if they want to alter text is to check the ID.

In future if API allows to ask for translations then there will be translations.

For the rest yes fixed is better long term.
If not then remotes should not have to query until the value is known at some point.

**Comment by Montellese** [comment]
@DaveTBlake: assuming we would always have a missing artist in the database (independent of whether its ID would be fixed or not) what are the implications for other parts of Kodi which might use the number of artists in the database as an indication for whether the library is empty or not (I don't know if that's the case but who knows)?

**Comment by DaveTBlake** [comment]
@Montellese count of artists is never used within Kodi (I can tell from Musicdatabase.cpp), but of course API users could do things with the artists they get, we have no way to know that. Kodi uses song count to determine if there is a music library, but good question.

What is your view on having a fixed and well known ID for this artificial artist? If implemented that way is there any need to add a GetMissingArtistID method to the API, can it just be documented instead?

**Comment by Montellese** [comment]
IMO the API must provide a way for the client to figure out what the ID of the "missing artist" is (independent of whether it's fixed or not). It could also be done as a property (like we already have for `Player` etc.) to choose a more generic approach in case we come across more special cases.

Personally I'd prefer a fixed ID (ideally the first possible ID) but that might be a problem for people who update from existing databases. But I don't really have a strong opinion on it.

**Comment by DaveTBlake** [comment]
@Montellese thanks for the second opinion I wanted, even if it is not a strong one.

I have changed to use a fixed artist id (=1) that always exists. Existing databases are updated to move record and related data in other tables if this ID is in use.  I think this is approached prefered by @Tolriq, but I'm sure he will tell us if not.

I will look at API change to provide a property that gives this missing tag artist ID. A generic approach sounds good, as I wonder if a similar approch to all this could be used for improvements to Various Artists (another PR for sure!!). Could API change be a separate PR follow up to this, or do I roll it in here?

**Comment by Tolriq** [comment]
You really see me way worse than I am ;)

Anyway IMO if API change comes later, I vote for a API version bump (even maybe just revision) so that API consumer knows that a new artist will be returned.

Revisions of version.txt is the way for any API consumer to know why some things change easily just by checking history. So just a change on it to have references to PR that modify behavior sounds correct.

**Comment by DaveTBlake** [comment]
@Montellese I have added API method to return the missing artist ID, if you would check that I have done this correctly and used suitable naming. I have taken the generic approach you suggested that allows for future extensions.

@Tolriq I have gone for API change too, I just want to get this completed and on to more interesting things. I can get a build going if you want to test it out, hardly recognisable since the early version you tried.

If everyone is happy I will then squish for merge.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 2]
```diff
@@ -41,6 +41,25 @@ using namespace JSONRPC;
...
 
+JSONRPC_STATUS CAudioLibrary::GetProperties(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  CVariant properties = CVariant(CVariant::VariantTypeObject);
+  for (unsigned int index = 0; index < parameterObject["properties"].size(); index++)
+  {
+    std::string propertyName = parameterObject["properties"][index].asString();
+    CVariant property;
+    if (propertyName == "missingartistid")
+    {
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp]: No curly braces on single line blocks.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 3]
```diff
@@ -41,6 +41,25 @@ using namespace JSONRPC;
 using namespace XFILE;
 using namespace KODI::MESSAGING;
 
+JSONRPC_STATUS CAudioLibrary::GetProperties(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  CVariant properties = CVariant(CVariant::VariantTypeObject);
+  for (unsigned int index = 0; index < parameterObject["properties"].size(); index++)
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp]: Please use iterators with `CVariant::begin_array()` and `CVariant::end_array()`.

File: xbmc/interfaces/json-rpc/schema/version.txt [hunk 4]
```diff
@@ -1 +1 @@
-7.8.1
+7.9.1
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/schema/version.txt]: 7.9.0

**Comment by Montellese** [comment]
JSON-RPC changes look ok apart from the minors.

**Comment by DaveTBlake** [comment]
Thanks @Montellese, minors fixed. 
Anything else anyone, or are we good to go?

**Comment by Tolriq** [comment]
Can't test before tomorrow but since the last version seems to address all the concerns and even performance of update all sounds OK.

**Comment by DaveTBlake** [comment]
Almost sounds hopeful;)
I've started a WIn32 test build, should be on the mirrors for testing in due course.

**Comment by Tolriq** [comment]
@MartijnKaijser I know not the correct place but it's important enough to be sure you have the message
Have something changed in the setup files of Kodi ? 
I started the test build installer from dave and my computer started a reboot (Windows 10) at the directx install phase.

I suppose the reboot was needed to finish directx install, but it should have asked me and not force reboot.

**Comment by MartijnKaijser** [comment]
NO this is not the correct place! You should know that by now!

**Comment by Tolriq** [comment]
Well I have proven that Trac & Forum are also not the correct places, until something works (And I'll gladly contribute to anything), at least you have the information, do whatever you want with it.

End of Off Topic.

About this, update went smooth in a few seconds and not 40 like first time :)
Missing artist is correctly added at 1, does not interfere with a [No artist] I already have + no impact on existing missing artist MBID.

Previous artist id 1 is correctly migrated will all data AFAICT.

display artist is empty and JSON works and return correct data.

No seen performance regression.

I cannot test the return of the text always in English as the code seems to support, since only English is present in the PR.

But a quick code review seems to indicate that a call to SetTranslateBlankArtist(false) is missing from JSON GetArtists no ? @DaveTBlake 

So except for the call to SetTranslateBlankArtist all seems to work perfectly.

**Comment by DaveTBlake** [comment]
Thanks for the testing and feedback @Tolriq 
It is staggaring how much indexes help the speed, well worth creating them even temporarily.

>But a quick code review seems to indicate that a call to SetTranslateBlankArtist(false) is missing from JSON GetArtists no ?

Yes sorry my omission, well spotted. Obviously subliminally I want to give API the translated value ;) A minor fix comming shortly.

**Comment by Tolriq** [comment]
Reading your last change opens a question for @Montellese.

It seems all access to db are references counted but JSON access to db mostly never close the database (Only in some cases in VideoLibrary). 
So leave a ref counter so db never close after a JSON access ?

**Comment by Montellese** [comment]
The DB connection is automatically closed by the destructor if it has not been done manually. This was not the case a few years ago but has been since then so if you see a manual call to close the connection it's old code or personal preference. The connection will be closed anyway.

**Comment by Tolriq** [comment]
Thanks, so there's no singleton or globals things patterns.

Since there's no more usage of the musicdatabase after the call at https://github.com/xbmc/xbmc/pull/9205/files#diff-37335a043764275ae506dce4f11d48f7R108

So there's no need to get previous translation status and set it back after. @DaveTBlake 

I was perplex to see those calls that implied a global state / ref counts, but then no close afterwards.

**Comment by DaveTBlake** [comment]
>So there's no need to get previous translation status and set it back after.

As yes, my mistake, don't know what I was thinking! Got that sorted.

jenkins build this please

**Comment by DaveTBlake** [comment]
Fix complier issues on other platforms
jenkins build this please.

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
Squished to 3 commits, keeping certain things separate. Build error was addon related, outside this PR.
Going to merge later today, unless someone has any objection.

**Comment by MilhouseVH** [comment]
Migration problem with MySQL... http://sprunge.us/PALL
```
18:49:39  19.471655 T:1962237952   ERROR: Unable to open database: MyMusic60 [1049](Unknown database 'MyMusic60')
18:49:39  19.515125 T:1962237952  NOTICE: Old database found - updating from version 59 to 60
18:49:44  25.156134 T:1962237952  NOTICE: Attempting to update the database MyMusic60 from version 59 to 60
18:49:44  25.156513 T:1962237952   DEBUG: Mysql Start transaction
18:49:45  26.260216 T:1962237952    INFO: UpdateTables - updating tables
18:49:45  26.260914 T:1962237952   DEBUG: Mysql execute: CREATE TEMPORARY TABLE tmp_artist CHARACTER SET utf8 COLLATE utf8_general_ci AS SELECT * FROM artist WHERE artist.idArtist = 1
18:49:45  26.275442 T:1962237952   DEBUG: Mysql execute: UPDATE tmp_artist SET idArtist = NULL
18:49:45  26.276043 T:1962237952   DEBUG: Mysql execute: INSERT INTO artist SELECT * FROM tmp_artist
18:49:45  26.276724 T:1962237952   DEBUG: Mysql execute: DROP TABLE tmp_artist
18:49:45  26.300764 T:1962237952   DEBUG: Mysql execute: DELETE FROM artist WHERE artist.idArtist = 1
18:49:45  26.301374 T:1962237952   DEBUG: Mysql execute: CREATE INDEX idxSongArtist2 ON song_artist ( idArtist )
18:49:46  26.482841 T:1962237952   DEBUG: Mysql execute: CREATE INDEX idxAlbumArtist2 ON album_artist ( idArtist )
18:49:46  26.538095 T:1962237952   DEBUG: Mysql execute: CREATE INDEX idxDiscography ON discography ( idArtist )
18:49:46  26.595396 T:1962237952   DEBUG: Mysql execute: CREATE INDEX ix_art ON art ( media_id, media_type(20) )
18:49:46  26.670818 T:1962237952   DEBUG: Mysql execute: UPDATE song_artist SET idArtist = 1553 WHERE idArtist = 1
18:49:46  26.674036 T:1962237952   DEBUG: Mysql execute: UPDATE album_artist SET idArtist = 1553 WHERE idArtist = 1
18:49:46  26.674936 T:1962237952   DEBUG: Mysql execute: UPDATE art SET media_id = 1553 WHERE media_id = 1 AND media_type='artist'
18:49:46  26.675537 T:1962237952   DEBUG: Mysql execute: UPDATE discography SET idArtist = 1553 WHERE idArtist = 1
18:49:46  26.676617 T:1962237952   DEBUG: Mysql execute: DROP INDEX idxSongArtist2
18:49:46  26.677017 T:1962237952   ERROR: SQL: [MyMusic60] Undefined MySQL error: Code (1064)
                                            Query: DROP INDEX idxSongArtist2
18:49:46  26.682171 T:1962237952   ERROR: Moving existing artist to add missing tag artist has failed
18:49:46  26.682339 T:1962237952   DEBUG: Mysql execute: INSERT INTO artist (idArtist, strArtist, strMusicBrainzArtistID) VALUES( 1, '[Missing Tag]', 'Artist Tag Missing' )
18:49:46  26.698336 T:1962237952   DEBUG: Mysql execute: CREATE INDEX idxSongArtist1 ON song_artist ( idSong, idRole )
18:49:46  26.866970 T:1962237952   DEBUG: Mysql execute: CREATE INDEX idxAlbumArtist1 ON album_artist ( idAlbum )
18:49:46  27.002516 T:1962237952   DEBUG: Mysql execute: DROP INDEX idxSongArtist1
18:49:46  27.002913 T:1962237952   ERROR: SQL: [MyMusic60] Undefined MySQL error: Code (1064)
                                            Query: DROP INDEX idxSongArtist1
18:49:46  27.003113 T:1962237952   ERROR: Exception updating database MyMusic60 from version 59 to 60
18:49:46  27.003143 T:1962237952   ERROR: Error updating database MyMusic60 from version 59 to 60
18:49:46  27.003866 T:1962237952   DEBUG: Mysql rollback transaction
18:49:46  27.033966 T:1962237952  NOTICE: Old database found - updating from version 58 to 60
18:49:46  27.036192 T:1962237952   ERROR: SQL: Can't create database for copy: 'MyMusic58' (1007)
18:49:46  27.036270 T:1962237952   ERROR: Unable to copy old database MyMusic58 to new version MyMusic60
```

**Comment by DaveTBlake** [comment]
Oh poop! MySQL not liking DROP INDEX statements, any ideas why @MilhouseVH ?

**Comment by MilhouseVH** [comment]
You need to specify the name of the table to which the index belongs - if SQLite allows otherwise that's probably something to avoid.

**Comment by zag2me** [comment]
Yep mysql needs DROP INDEX index_name ON tbl_name I think

**Comment by DaveTBlake** [comment]
Thanks guys, caught out by MySQL again :(
How best to handle this? Revert the PR or just raise another to fix it? I would prefer latter, makes it more obvious what the issue are, and I have already rebased other work following the merge. I have another MySQL issue to fix too.

**Comment by MilhouseVH** [comment]
I'd just push another PR that fixes the `DROP INDEX` issues - presumably it will be OK to keep the version number the same as long as the new PR goes in quickly.

**Comment by MilhouseVH** [comment]
I'm pretty sure SQLite supports the same syntax as MySQL - see here for existing code that drops an index: https://github.com/xbmc/xbmc/blob/master/xbmc/video/VideoDatabase.cpp#L4805

**Comment by MilhouseVH** [comment]
If `exec()` handles it, I'd use that.

**Comment by DaveTBlake** [comment]
But documentation says MySQL needs
`ALTER TABLE tablename DROP INDEX indexname`
so does the `DROP INDEX indexname ON tablename` work OK or is there another hidden trick?

**Comment by Montellese** [comment]
Please read my previous message. Whenever you update the database version we drop all views, indices etc. and re-create them. So you don't have to manually drop it. You just need to remove that line from `CreateAnalytics()`.

**Comment by DaveTBlake** [comment]
>Please read my previous message. Whenever you update the database version we drop all views, indices etc. and re-create them. So you don't have to manually drop it. You just need to remove that line from CreateAnalytics().

@Montellese I read and understand the update process, but what I am doing is necessary. 
UpdateTables runs after the old DB has been copied to the new and the indexes dropped. In change from 59 to 60 the user data (not just the data structure) needs to be checked and modified. UpdateTables needs to select all songs that do not have an artist and update them, and move all data that belongs to artist ID = 1. Doing this without indices is very slow, hence I temporarily create just the indexes needed to make these data changes efficiently. At the end of the process I drop them. Some will subsequently be recreated by `Create_Analytics`, but since the data update and manual index creation does not always happen it is cleaner to do this.
